From 0eb666752a7e8eba3aff218a5165b651e5dc8837 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Thu, 10 Jun 2021 15:15:48 +0200
Subject: [PATCH] PONRTSYS-9035: Allocate device specific CMA memory

Reserve device specific CMA memory if a memory-region attribute was
added to the device tree. This allows to specify an additional CMA
region which is only used by the buffer manager.

This functionality is optional, if it is not specified the global CMA
memory will be used.
---
 drivers/net/ethernet/lantiq/cqm/Kconfig      |  1 +
 drivers/net/ethernet/lantiq/cqm/prx300/cqm.c | 13 +++++++++++++
 2 files changed, 14 insertions(+)

diff --git a/drivers/net/ethernet/lantiq/cqm/Kconfig b/drivers/net/ethernet/lantiq/cqm/Kconfig
index 3ab8140f59b4..cd7ab390d47b 100644
--- a/drivers/net/ethernet/lantiq/cqm/Kconfig
+++ b/drivers/net/ethernet/lantiq/cqm/Kconfig
@@ -17,6 +17,7 @@ choice
 
 config PRX300_CQM
 	bool "PRX300"
+	select OF_RESERVED_MEM
 	help
 	  CQM driver for PRX300 platform.
 if PRX300_CQM
diff --git a/drivers/net/ethernet/lantiq/cqm/prx300/cqm.c b/drivers/net/ethernet/lantiq/cqm/prx300/cqm.c
index 5c6c64907c6e..bd3f524fe1d3 100644
--- a/drivers/net/ethernet/lantiq/cqm/prx300/cqm.c
+++ b/drivers/net/ethernet/lantiq/cqm/prx300/cqm.c
@@ -1,6 +1,7 @@
 #include "cqm.h"
 #include "cqm_debugfs.h"
 #include "../cqm_dev.h"
+#include <linux/of_reserved_mem.h>
 #include <net/datapath_proc_api.h>
 #include <net/switch_api/gsw_flow_ops.h>
 #include <net/switch_api/gsw_dev.h>
@@ -4931,6 +4932,17 @@ static int cqm_prx300_probe(struct platform_device *pdev)
 			return -ENOENT;
 		}
 	}
+
+	ret = of_reserved_mem_device_init_by_idx(&pdev->dev, pdev->dev.of_node,
+						 0);
+	/* Ignore if this is not available in the device tree, we will use
+	 * the normal CMA memory instead.
+	 */
+	if (ret && ret != -ENODEV) {
+		dev_err(&pdev->dev, "Error adding reserved mem: %i\n", ret);
+		return ret;
+	}
+
 	#endif
 
 	cqm_ctrl->name = cqm_name;
@@ -5109,6 +5121,7 @@ static int cqm_prx300_probe(struct platform_device *pdev)
 
 static int cqm_prx300_release(struct platform_device *pdev)
 {
+	of_reserved_mem_device_release(&pdev->dev);
 	return 0;
 }
 
